# Assignment 5, Question 4: Data Exploration

**Points: 15**

In this notebook, you'll explore the clinical trial dataset using pandas selection and filtering techniques.

You'll use utility functions from `q3_data_utils` where helpful, but also demonstrate direct pandas operations.

## Setup

In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# Import utilities from Q3
from q3_data_utils import load_data, detect_missing, filter_data

# Load the data
df = load_data('data/clinical_trial_raw.csv')
print(f"Loaded {len(df)} patients with {len(df.columns)} variables")

# Prewritten visualization functions for exploration
def plot_value_counts(series, title, figsize=(10, 6)):
    """
    Create a bar chart of value counts.
    
    Args:
        series: pandas Series with value counts
        title: Chart title
        figsize: Figure size tuple
    """
    plt.figure(figsize=figsize)
    series.plot(kind='bar')
    plt.title(title)
    plt.xticks(rotation=45)
    plt.tight_layout()
    plt.show()

def plot_crosstab(crosstab_data, title, figsize=(10, 6)):
    """
    Create a heatmap of crosstab data.
    
    Args:
        crosstab_data: pandas DataFrame from pd.crosstab()
        title: Chart title
        figsize: Figure size tuple
    """
    plt.figure(figsize=figsize)
    plt.imshow(crosstab_data.values, cmap='Blues', aspect='auto')
    plt.colorbar()
    plt.title(title)
    plt.xticks(range(len(crosstab_data.columns)), crosstab_data.columns, rotation=45)
    plt.yticks(range(len(crosstab_data.index)), crosstab_data.index)
    plt.tight_layout()
    plt.show()

'Data shape: (10000, 18)'

Loaded 10000 patients with 18 variables


## Part 1: Basic Exploration (3 points)

Display:
1. Dataset shape
2. Column names and types
3. First 10 rows
4. Summary statistics (.describe())

In [6]:
# TODO: Display dataset info
display(df.shape)
df.columns
df.dtypes
df.head(n=10)
df.describe()

(10000, 18)

,age,bmi,systolic_bp,diastolic_bp,cholesterol_total,cholesterol_hdl,cholesterol_ldl,glucose_fasting,follow_up_months,adverse_events,adherence_pct
count,10000.000000,9562.000000,9586.000000,9586.000000,9446.000000,9446.000000,9446.000000,9631.000000,10000.00000,10000.000000,8533.000000
mean,59.182700,25.730558,117.531087,73.550908,178.039488,61.369786,85.698603,96.424255,12.25460,0.145500,60.615610
std,151.769963,5.339547,13.973973,10.167464,33.129034,11.062101,28.686463,17.112961,7.07675,0.393631,18.974399
min,-999.000000,-1.000000,75.000000,60.000000,91.000000,25.000000,40.000000,51.000000,0.00000,0.000000,20.000000
25%,70.000000,23.500000,108.000000,65.000000,155.000000,54.000000,65.000000,84.000000,6.00000,0.000000,47.000000
50%,80.000000,26.000000,117.000000,73.000000,177.000000,61.000000,84.000000,96.000000,12.00000,0.000000,62.000000
75%,92.000000,28.775000,127.000000,81.000000,200.000000,69.000000,105.000000,108.000000,19.00000,0.000000,75.000000
max,100.000000,42.800000,173.000000,118.000000,315.000000,98.000000,226.000000,163.000000,24.00000,4.000000,100.000000


## Part 2: Column Selection (3 points)

Demonstrate different selection methods:

1. Select only numeric columns using `.select_dtypes()`
2. Select specific columns by name
3. Select a subset of rows and columns using `.loc[]`

In [7]:
# TODO: Select numeric columns
numeric_df = df.select_dtypes(include=['number'])
display(numeric_df.columns)

Index(['age', 'bmi', 'systolic_bp', 'diastolic_bp', 'cholesterol_total',
       'cholesterol_hdl', 'cholesterol_ldl', 'glucose_fasting',
       'follow_up_months', 'adverse_events', 'adherence_pct'],
      dtype='object')

In [10]:
# TODO: Select specific columns
df[['age', 'cholesterol_total', 'bmi']]

,age,cholesterol_total,bmi
0,80,120.0,29.3
1,80,206.0,NaN
2,82,172.0,-1.0
3,95,200.0,25.4
4,95,185.0,NaN
...,...,...,...
9995,72,182.0,23.2
9996,100,157.0,28.9
9997,78,154.0,23.8
9998,86,196.0,27.0


In [13]:
# TODO: Use .loc[] to select subset
display(df.loc[0:3, ['age', 'cholesterol_total']])  

,age,cholesterol_total
0,80,120.0
1,80,206.0
2,82,172.0
3,95,200.0


## Part 3: Filtering (4 points)

Filter the data to answer these questions:

1. How many patients are over 65 years old?
2. How many patients have systolic BP > 140?
3. Find patients who are both over 65 AND have systolic BP > 140
4. Find patients from Site A or Site B using `.isin()`

In [2]:
# TODO: Filter and count patients over 65
# 1. Use the filter_data utility from Q3
test_filters = [{'column': 'age', 'condition': 'greater_than', 'value': 65}]
# 2. Create a filter for age > 65
patients_over_65 = filter_data(df, test_filters)
# 3. Apply the filter and count the results
print(f"Patients over 65: {len(list(patients_over_65))}")


Patients over 65: 18


In [17]:
# TODO: Filter for high BP
# 1. Use the filter_data utility from Q3
# 2. Create a filter for systolic_bp > 140
high_bp_filters = [{'column': 'systolic_bp', 'condition': 'greater_than', 'value': 140}]
# 3. Apply the filter and count the results
high_bp = filter_data(df, high_bp_filters)
# print(f"Patients with high BP: {len(high_bp)}")
print(f"Patients with high BP: {len(list(high_bp))}")

Patients with high BP: 18


In [18]:
# TODO: Multiple conditions with &
# 1. Use filter_data for multiple conditions:
# 2. Create filters for both conditions:
#     {'column': 'age', 'condition': 'greater_than', 'value': 65},
#     {'column': 'systolic_bp', 'condition': 'greater_than', 'value': 140}
# ]
filter_age_and_bp = [
    {'column': 'age', 'condition': 'greater_than', 'value': 65},
    {'column': 'systolic_bp', 'condition': 'greater_than', 'value': 140}
]
# 3. Apply the filter and count the results
both_conditions = filter_data(df, filter_age_and_bp)
print(f"Patients over 65 AND high BP: {len(both_conditions)}")
# 5. Alternative: Use in_range for age range:
# 5. Create filter for age range 65-100
# 6. Apply the filter and count the results


Patients over 65 AND high BP: 0


In [23]:
# TODO: Filter by site using .isin()
# 1. Use the filter_data utility from Q3
# 2. Create a filter for Site A or Site B
filter_site_ab = df[df['site'].isin(['Site A', 'Site B'])]
print("\\nSite A or Site B:")
print(filter_site_ab)
# 3. Apply the filter and count the results
site_ab = filter_data(df, filter_site_ab)
# print(f"Patients from Site A or B: {len(site_ab)}")
print(f"Patients from Site A or B: {len(list(site_ab))}")
print(f"Patients from Site A or B: {len(site_ab)}")

\nSite A or Site B:
     patient_id  age         sex   bmi enrollment_date  systolic_bp  \
1        P00002   80    Female     NaN      2022-01-06        139.0   
5        P00006   78           F  26.8      2023-08-29        116.0   
10       P00011   79      Female  23.5      2023-06-12        110.0   
18       P00019  100           M  24.6      2022-01-17        115.0   
20       P00021   70      Female  28.9      03/18/2022        131.0   
...         ...  ...         ...   ...             ...          ...   
9954     P09955   77      Female  24.4      2022-08-11        115.0   
9969     P09970   85           F  22.2      2023-06-24        111.0   
9978     P09979   66           F  25.9      2023-07-02        124.0   
9986     P09987   96           M  27.7      2022-05-04        104.0   
9999     P10000   67      Female  29.4      25-03-2022        134.0   

      diastolic_bp  cholesterol_total  cholesterol_hdl  cholesterol_ldl  \
1             81.0              206.0             58

## Part 4: Value Counts and Grouping (5 points)

1. Get value counts for the 'site' column
2. Get value counts for the 'intervention_group' column  
3. Create a crosstab of site vs intervention_group
4. Calculate mean age by site
5. Save the site value counts to `output/q4_site_counts.csv`

In [4]:
# TODO: Value counts and analysis
def count_site_values(df):
    site_counts = df['site'].value_counts()
    return site_counts
print(f"\\nSite counts: {count_site_values(df)}")

#Value counts for intervention group[
def count_intervention_values(df):
    intervention_counts = df['intervention_group'].value_counts()
    return intervention_counts
print(f"\\nIntervention group counts: {count_intervention_values(df)}")

#Cross-tab of site vs intervention group
pd.pivot_table(df, values='patient_id', index='intervention_group', columns='site', aggfunc='count')

#Calculate mean age by site
pd.pivot_table(df, values='age', index='site', aggfunc='mean')

\nSite counts: site
site b         742
Site B         736
SITE B         703
SITE A         684
Site  A        681
Site A         661
Site C         658
site a         651
site c         615
SITE C         605
Site D         362
site d         349
Site_D         332
Site E         319
SITE D         313
SITE E         295
site e         294
  SITE B        94
  site b        90
  Site B        88
  Site C        83
  site a        74
  SITE A        74
  Site  A       67
  Site A        64
  site c        57
  SITE C        55
  Site E        42
  SITE D        41
  site d        41
  site e        36
  Site D        32
  SITE E        31
  Site_D        31
Name: count, dtype: int64
\nIntervention group counts: intervention_group
Contrl              802
TREATMENT B         761
Treatment  B        760
Control             751
treatment b         750
control             734
Treatment B         730
CONTROL             715
TreatmentA          635
Treatment A         610
Treatmen A          

,age
site,
SITE A,23.608108
SITE B,59.670213
SITE C,60.254545
SITE D,77.853659
SITE E,12.258065
Site A,80.328358
Site A,49.843750
Site B,56.318182
Site C,27.831325


In [36]:
# TODO: Save output
# site_counts.to_csv('output/q4_site_counts.csv')
count_site_values(df).to_csv('output/q4_site_counts.csv')

## Summary

Write 2-3 sentences about what you learned from exploring this dataset.

**Your summary here:**

TODO: Write your observations


The mean age of 10,000 participants in this sample is 59.18, with 18 participants over the age of 65. Participants' systolic blood pressure ranges from [75, 173]; only 18 participants have high systolic blood pressure ranging from [140, 173]. There are no participants that were over the age of 65 and had high systolic blood pressure. 